packages/optimizer/lib/transformers/OptimizeHeroImages.js

```diff
@@ -0,0 +1,349 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const {
+  appendChild,
+  createElement,
+  hasAttribute,
+  insertAfter,
+  nextNode,
+  firstChildByTag,
+} = require('../NodeUtils');
+const {findMetaViewport, skipNodeAndChildren} = require('../HtmlDomHelper');
+const {isValidImageSrcURL} = require('../URLUtils');
+const {isTemplate, isAmpStory} = require('../AmpConstants');
+
+// Images smaller than 150px are considered tiny
+const TINY_IMG_THRESHOLD = 150;
+// Maximum number of hero images defined via data-hero
+const DATA_HERO_MAX = 2;
+
+/**
+ * OptimizeHeroImage - this transformers optimizes image rendering times for hero images. For hero
+ * images it will:
+ *
+ * 1. Inject a preload hint (if possible)
+ * 2. Generate an img tag enabling the browser to render the image without the AMP runtime being loaded.
+ *
+ * Hero images are either identified automatically or can be explicitly defined by adding an `data-hero`
+ * attribute to the element.
+ *
+ * This transformer supports the following options:
+ *
+ * * `optimizeHeroImages`: [true|false] - enables or disables hero image optimization. The default is `true`.
+ * * `maxHeroImageCount`: [int] - defines the maximum number of hero images per page. The default is `2`.
+ */
+class OptimizeHeroImage {
+  constructor(config) {
+    this.log = config.log;
+    this.enabled = config.optimizeHeroImages !== false || config.preloadHeroImage !== false;
```

Comment:
should a warning be emitted about using `preloadHeroImage`?

Reply:
done